File: src/main/java/fr/adrienbrault/idea/symfony2plugin/routing/inspection/TwigRouteMissingInspection.java
package fr.adrienbrault.idea.symfony2plugin.routing.inspection;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.patterns.ElementPattern;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.templating.TwigPattern;
import fr.adrienbrault.idea.symfony2plugin.routing.RouteHelper;
import fr.adrienbrault.idea.symfony2plugin.templating.util.TwigUtil;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class TwigRouteMissingInspection extends LocalInspectionTool {
    @NotNull
    public PsiElementVisitor buildVisitor(final @NotNull ProblemsHolder holder, boolean isOnTheFly) {
        if(!Symfony2ProjectComponent.isEnabled(holder.getProject())) {
            return super.buildVisitor(holder, isOnTheFly);
        }

        return new MyPsiElementVisitor(holder);
    }

    private static class MyPsiElementVisitor extends PsiElementVisitor {
        @NotNull private final ProblemsHolder holder;
        private ElementPattern<?> autocompletableRoutePattern;

        MyPsiElementVisitor(@NotNull ProblemsHolder holder) {
            this.holder = holder;
        }

        @Override
        public void visitElement(@NotNull PsiElement element) {
            if(getAutocompletableRoutePattern().accepts(element) && TwigUtil.isValidStringWithoutInterpolatedOrConcat(element)) {
                invoke(element, holder);
            }

            super.visitElement(element);
        }

        private ElementPattern<?> getAutocompletableRoutePattern() {
            return autocompletableRoutePattern != null ? autocompletableRoutePattern : (autocompletableRoutePattern = TwigPattern.getAutocompletableRoutePattern());
        }
    }

    private static void invoke(@NotNull final PsiElement element, @NotNull ProblemsHolder holder) {
        String text = element.getText();
        if (StringUtils.isBlank(text)) {
            return;
        }

        if (!RouteHelper.isExistingRouteName(element.getProject(), RouteHelper.unescapeRouteName(text))) {
            holder.registerProblem(element, "Symfony: Missing Route", new RouteGuessTypoQuickFix(text));
        }
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/routing/inspection/TwigRouteMissingInspection.kt
package fr.adrienbrault.idea.symfony2plugin.routing.inspection

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.patterns.ElementPattern
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.routing.RouteHelper
import fr.adrienbrault.idea.symfony2plugin.templating.TwigPattern
import fr.adrienbrault.idea.symfony2plugin.templating.util.TwigUtil
import org.apache.commons.lang3.StringUtils

private fun inspectTwigRouteMissing(element: PsiElement, holder: ProblemsHolder) {
    val text = element.text
    if (StringUtils.isBlank(text)) {
        return
    }

    val routeName = RouteHelper.unescapeRouteName(text)
    if (!RouteHelper.isExistingRouteName(element.project, routeName)) {
        holder.registerProblem(element, "Symfony: Missing Route", RouteGuessTypoQuickFix(text))
        return
    }

    if (RouteHelper.isRouteControllerDeprecated(element.project, routeName)) {
        holder.registerProblem(element, "Symfony: Controller action is deprecated", ProblemHighlightType.LIKE_DEPRECATED)
    }
}

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class TwigRouteMissingInspection : LocalInspectionTool() {
    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        if (!Symfony2ProjectComponent.isEnabled(holder.project)) {
            return super.buildVisitor(holder, isOnTheFly)
        }

        return MyPsiElementVisitor(holder)
    }

    private class MyPsiElementVisitor(private val holder: ProblemsHolder) : PsiElementVisitor() {
        private var autocompletableRoutePattern: ElementPattern<*>? = null

        override fun visitElement(element: PsiElement) {
            if (getAutocompletableRoutePattern().accepts(element) && TwigUtil.isValidStringWithoutInterpolatedOrConcat(element)) {
                inspectTwigRouteMissing(element, holder)
            }

            super.visitElement(element)
        }

        private fun getAutocompletableRoutePattern(): ElementPattern<*> {
            if (autocompletableRoutePattern != null) {
                return autocompletableRoutePattern!!
            }

            autocompletableRoutePattern = TwigPattern.getAutocompletableRoutePattern()
            return autocompletableRoutePattern!!
        }
    }
}
